src/p5.dance.js

```diff
@@ -183,6 +183,46 @@ module.exports = class DanceParty {
     this.allSpritesLoaded = true;
   }
 
+  setUserBlocks(userBlocks) {
+    this.userBlocksWithNextBlock = [];
+    userBlocks.forEach(b => {
+      var block = {type: b.type, id: b.id};
+      if (b.childBlocks_[0]) {
+        block.nextBlockType = b.childBlocks_[0].type;
+        block.nextBlockId = b.childBlocks_[0].id;
+      }
+      this.userBlocksWithNextBlock.push(block);
+    });
+  }
+
+  getUserBlocks() {
+    return this.userBlocksWithNextBlock.map(b => b.type);
+  }
+
+  getUserBlocksWithNextBlock() {
+    return this.userBlocksWithNextBlock;
+  }
+
+  // This function checks if there is an AI block anywhere in workspace.
+  // Note that the AI block could be disconnected and technically not part of the user code.
+  hasAiBlock() {
+    return this.getUserBlocks().includes('Dancelab_ai');
+  }
+
+  // This function checks if the AI block is one of the children blocks of the given 'top' block.
+  // For example, if the given block is a 'When up pressed' event block, this function checks
+  // if the AI block is included within the event block.
+  isAiBlockChildOfTopBlock(topBlockType) {
+    let topBlock = this.userBlocksWithNextBlock.find(b => b.type === topBlockType);
+    while (topBlock.nextBlockType) {
+      if (topBlock.nextBlockType === 'Dancelab_ai') {
+        return true;
+      }
+      topBlock = this.userBlocksWithNextBlock.find(b => b.id == topBlock.nextBlockId);
```

Comment:
Does this represent a `topBlock` when set here? Or is it just we're iterating through each of the blocks below the top block initialized in the first line of this function? If so, might be more clear to initialize `topBlock` with the first child of the top block type being passed in, and rename to just `block` or something.

Reply:
Yes, `topBlock` does represent the to block when initialized, but then iterating through the blocks below it. I'll define a new variable called `block` for the iteration. Thanks!

Reply:
I was having a conversation with myself about this, so thanks for airing it!